perllib/FixMyStreet/App/Controller/Admin.pm

```diff
@@ -514,35 +574,94 @@ sub lookup_body : Private {
     return 1;
 }
 
+sub body_base : Chained('/') : PathPart('admin/body') : CaptureArgs(0) { }
+
 # This is for if the category name contains a '/'
-sub category_edit_all : Path('body') {
+sub category_edit_all : Chained('body_base') : PathPart('') {
     my ( $self, $c, $body_id, @category ) = @_;
     my $category = join( '/', @category );
-    $c->go( 'category_edit', [ $body_id, $category ] );
-}
 
-sub category_edit : Path('body') : Args(2) {
-    my ( $self, $c, $body_id, $category ) = @_;
+    $c->stash->{body_id} = $body_id;
+    $c->forward( 'lookup_body' );
 
+    my $contact = $c->stash->{body}->contacts->search( { category => $category } )->first;
+    $c->stash->{contact} = $contact;
+
+    $c->stash->{template} = 'admin/category_edit.html';
+    $c->forward( 'category_edit' );
+}
+
+sub body : Chained('body_base') : PathPart('') : CaptureArgs(1) {
+    my ( $self, $c, $body_id ) = @_;
     $c->stash->{body_id} = $body_id;
+}
+
+sub edit_body : Chained('body') : PathPart('') : Args(0) {
+    my ( $self, $c ) = @_;
+
+    unless ($c->user->has_permission_to('category_edit', $c->stash->{body_id})) {
+        $c->forward('check_for_super_user');
+    }
+
+    $c->forward( '/auth/get_csrf_token' );
+    $c->forward( 'lookup_body' );
+    $c->forward( 'fetch_all_bodies' );
+    $c->forward( 'body_form_dropdowns' );
+    $c->forward('fetch_languages');
+
+    if ( $c->get_param('posted') ) {
+        $c->forward('update_contacts');
+    }
+
+    $c->stash->{object} = $c->stash->{body};
+    $c->stash->{translation_col} = 'name';
+
+    # if there's a contact then it's becuase we're displaying error
+    # messages about adding a contact so grabbing translations will
+    # fetch the contact submitted translations. So grab them, stash
+    # them and then clear posted so we can fetch the body translations
+    if ($c->stash->{contact}) {
```

Comment:
Sorry, this still doesn't work. `$c->stash->{contact}` is not set when a category is successfully created or edited, because that submission comes to this function, `edit_body`, so the category translations are still then shown in the edit body form you get. I see what this is doing if there is an error, but I think you need to set posted to false regardless. (At least I think that would work - I think this also needs a new test that upon submitting a new category/editing a category, the body translations are still shown in the right place :) )